### PR TITLE
Remove support for Netscape certs

### PR DESCRIPTION
**Pull request by richsalz**
This removes support for the old Netscape cert extensions.

To quote email from Mozilla folks I have: "In general if there is an IETF PKIX replacement for a Netscape datatype, then you don't need to support the Netscape datatype. Therefore OpenSSL does not need to support Netscape cert types and Netscape cert sequence."

This is the first of a couple planned PR's in this area.

### CODE REVIEW

**Comment by richsalz** [comment]
There is a build failure:
```
    # ERROR: (int) 'EC_GROUP_cmp(group, group2, NULL) == 0' failed @ test/ectest.c:1826
    # [1] compared to [0]
    not ok 1 - parameter_test
```
but that also fails on master to not relevant here.

**Comment by t-j-h** [comment]
-1 blocker to enable discussion to occur before approval

File: crypto/x509/v3_purp.c [hunk 1]
```diff
@@ -644,19 +620,7 @@ static int purpose_smime(const X509 *x, int ca)
...
-            return 0;
-    }
-    if (x->ex_flags & EXFLAG_NSCERT) {
-        if (x->ex_nscert & NS_SMIME)
-            return 1;
-        /* Workaround for some buggy certificates */
-        if (x->ex_nscert & NS_SSL_CLIENT)
-            return 2;
-        return 0;
+        return ca_ret;
```
Comment (tmshort) [inline crypto/x509/v3_purp.c]: The two lines above this are redundant with this line.
However, the original code would return 0 if ca_ret = 5.

File: crypto/x509/v3_purp.c [hunk 2]
```diff
@@ -570,14 +556,9 @@ int X509_check_ca(X509 *x)
...
+
     ca_ret = check_ca(x);
-    if (!ca_ret)
-        return 0;
-    /* check nsCertType if present */
-    if (ca_ret != 5 || x->ex_nscert & NS_SSL_CA)
-        return ca_ret;
-    else
-        return 0;
+    return ca_ret;
```
Comment (tmshort) [inline crypto/x509/v3_purp.c]: This function could just be:
```
return check_ca(x);
```
But, then with this simplification, the `check_ssl_ca()` function could be replaced with `check_ca()`

**Comment by richsalz** [comment]
additional commit pushed, @tmshort please take a look.

File: crypto/x509/v3_purp.c [hunk 3]
```diff
@@ -625,9 +587,10 @@ static int check_purpose_ns_ssl_server(const X509_PURPOSE *xp, const X509 *x,
                                        int ca)
 {
     int ret;
+
     ret = check_purpose_ssl_server(xp, x, ca);
     if (!ret || ca)
-        return ret;
+        return 0;
```
Comment (tmshort) [inline crypto/x509/v3_purp.c]: Is it possible that both `ret` and `ca` could be non-zero? In such a case, the return value would be non-zero.  The same thing at L617 and L682, below.

Comment (richsalz) [reply crypto/x509/v3_purp.c]: I think it's not possible, but I'm not positive, to I reverted to ```return ret``` and updated the commit.  good point, thanks.

**Comment by richsalz** [comment]
Thoughts on this?  Netscape cert extensions should really be removed; even the maintainers of NSS think so. :)

**Comment by levitte** [comment]
I can't remember seeing much of a discussion related to this.  It's been a long time since I've seen the netscape extensions, so I can't say that I'd miss much of anything that's removed in this PR.

Considering none of them are critical (as far as I know), and they have been deprecated for a long time (see the docs), I'm not sure I understand the hold.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> I can't remember seeing much of a discussion related to this.  It's been a long time since I've seen the netscape extensions, so I can't say that I'd miss much of anything that's removed in this PR.
>
> Considering none of them are critical (as far as I know), and they have been deprecated for a long time (see the docs), I'm not sure I understand the hold.
>

In certificate list on my linux I cound find 2 certificates - one 
expired this month another one in 2027 - Issuer: C=FR, O=Dhimyotis, 
CN=Certigna

So why to remove them ?

**Comment by richsalz** [comment]
The Certigna cert is here: https://crt.sh/?id=62358

Look at the details of the cert – it’s a root CA – and the netscape extensions are *WRONG* but the PKIX extensions are *RIGHT*

**Comment by levitte** [comment]
Not sure that the NS extensions are actually wrong...  all they say is that this is a CA cert, and that it can be used in all kinds of contexts: SSL, S/MIME or object signing.  In other words, it can be used to verify leaf certificates with all those uses.  Or have I misunderstood NS extensions?

**Comment by levitte** [comment]
That being said, since the PKIX extensions are in place, serving the same purpose, *and the key usage marked critical* (as it should), the NS extensions really don't serve much of a purpose, except commentary.

**Comment by richsalz** [comment]
You're not quite reading it correctly; the NS extensions say it can sign other specific types of CA's. But you are correct about the critical keyUsage extension. So I think that while @petrovr pointed out an interesting cert, it won't be impacted if this PR is merged.

File: crypto/x509/ext_dat.h [hunk 4]
```diff
@@ -9,7 +9,7 @@
 
 int name_cmp(const char *name, const char *cmp);
 
-extern const X509V3_EXT_METHOD v3_bcons, v3_nscert, v3_key_usage, v3_ext_ku;
+extern const X509V3_EXT_METHOD v3_bcons, v3_key_usage, v3_ext_ku;
 extern const X509V3_EXT_METHOD v3_pkey_usage_period, v3_sxnet, v3_info, v3_sinfo;
 extern const X509V3_EXT_METHOD v3_ns_ia5_list[8], v3_alt[3], v3_skey_id, v3_akey_id;
```
Comment (FdaSilvaYY) [inline crypto/x509/ext_dat.h]: `v3_ns_ia5_list` should be removed too

File: crypto/x509/standard_exts.h [hunk 5]
```diff
@@ -13,14 +13,6 @@
  */
 
 static const X509V3_EXT_METHOD *standard_exts[] = {
-    &v3_nscert,
-    &v3_ns_ia5_list[0],
```
Comment (FdaSilvaYY) [inline crypto/x509/standard_exts.h]: `v3_ia5.c` file should be removed too, as nothing more is using `v3_ns_ia5_list`

File: crypto/x509/v3_bitst.c [hunk 6]
```diff
@@ -13,18 +13,6 @@
...
-    {2, "S/MIME", "email"},
-    {3, "Object Signing", "objsign"},
-    {4, "Unused", "reserved"},
-    {5, "SSL CA", "sslCA"},
-    {6, "S/MIME CA", "emailCA"},
-    {7, "Object Signing CA", "objCA"},
-    {-1, NULL, NULL}
-};
-
 static BIT_STRING_BITNAME key_usage_type_table[] = {
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_bitst.c]: `static const` while you are here

File: include/openssl/x509v3.h [hunk 7]
```diff
@@ -350,7 +350,7 @@ struct ISSUING_DIST_POINT_st {
 # define EXFLAG_BCONS            0x1
 # define EXFLAG_KUSAGE           0x2
 # define EXFLAG_XKUSAGE          0x4
-# define EXFLAG_NSCERT           0x8
+# define EXFLAG_NSCERT           0x0
```
Comment (FdaSilvaYY) [inline include/openssl/x509v3.h]: Add a note to say that `its old value should not reused` ?

Comment (richsalz) [reply include/openssl/x509v3.h]: done.

File: crypto/x509/v3_bitst.c (hunk 6 above)
Comment (richsalz) [reply crypto/x509/v3_bitst.c]: can't, because of use in the macro below; added a comment.

File: crypto/x509/standard_exts.h (hunk 5 above)
Comment (richsalz) [reply crypto/x509/standard_exts.h]: ooh, cool.  removed the file.

File: crypto/x509/ext_dat.h (hunk 4 above)
Comment (richsalz) [reply crypto/x509/ext_dat.h]: done.

**Comment by richsalz** [comment]
I think in view of the commit I just pushed, this still needs another review.

**Comment by FdaSilvaYY** [comment]
Oups, something disappears :
```
# Number of lines in @def_lines before massaging: 4635
# Number of lines in @nm_lines after massaging: 4629
# Number of lines in @def_lines after massaging: 4631
# The following symbols are missing in libcrypto.so:
#   i2s_ASN1_IA5STRING
#   s2i_ASN1_IA5STRING
not ok 2 - check that there are no missing symbols in libcrypto.so

```

**Comment by richsalz** [comment]
Updated commit pushed.

**Comment by richsalz** [comment]
Should the i2s/s2i_IA5STRING functions be removed?

**Comment by richsalz** [comment]
To answer my question: no, the functions should not be removed.  I put them back, but in another file.

Tim put a -1/hold on this three weeks ago. What's the next step?

**Comment by t-j-h** [comment]
If someone within the OMC disagrees with the view, then they call a vote to get a formal OMC decision. If no-one in the OMC wants to take it to a vote then it remains blocked.

I disagree that the maintainers of NSS think we should *remove* this from OpenSSL. They might think they don't currently use it themselves - and they might recommend using PKIX I'm preference - but that is different from us deciding to *remove* the code. I think you have missed that context.

And yes I did actually talked to a key NSS person about the distinction between the two contexts.

**Comment by paulidale** [comment]
With @levitte having already approved this, a vote might be desirable.

**Comment by petrovr** [comment]
In this issue we could see clean up of openssl configuration. This part is fine - if extension is not used then exclude from sample configuration.

Another part is to remove support from code even without deprecation period! I know projects where some functionality still exist even if it was deprecated more then 15 years ago.

So what about to mark related "definitions NS_*" x509v3.h as deprecated and in 2028 to remove them?

**Comment by richsalz** [comment]
@t-j-h I also spoke with a couple of "key" RedHat and Mozilla folks and got a different take.

The documentation has marked the extensions as deprecated for more than a decade:
```
; git blame doc/man5/x509v3_config.pod | grep -i netscape
5dd87981 doc/apps/x509v3_config.pod (Dr. Stephen Henson 2004-11-17 00:55:43 +0000 424) The following extensions are non standard, Netscape specific and largely
19f39703 doc/apps/x509v3_config.pod (Dr. Stephen Henson 2004-11-16 14:09:12 +000
```

Isn't that long enough?

**Comment by kroeckx** [comment]
There probably is still software that supports this. I think we
removed something simular in 1.1.0 and someone complained that
their product still supported importing such certificates and was
now broken.

**Comment by t-j-h** [comment]
@richsalz did you ask them if they can **guarantee** that there are no active users of this capability at all?

That is the **right question** to be asked for compatibility and matches our context.
And the previous stuff you forwarded in this PR is answering a rather different question. 

Anyone making an assertion that they don't exist and you cannot find them in usage anywhere would be wrong - as already noted in a single instance in this PR.

And that capability was not marked as **deprecated** in the **code**. 
The docs also noted explicitly 

> Their use in new applications is **discouraged**.

**Comment by richsalz** [comment]
No I did not ask if they can guarantee, the answer would be as ridiculous as asking if OpenSSL could gaurantee any kind of use.

Yes, the instance was reported, and I commented on it.

This has a hold and two approvals, it seems that the right thing is for the OMC to hold a discussion and vote.  If nothing happens in a couple of weeks I'll close the PR and delete the branch.

**Comment by t-j-h** [comment]
@richsalz it has one OMC hold, one OMC approval, and one OMC questioning if this should actually be removed (public PR comments). It is being discussed.

**Comment by richsalz** [comment]
@t-j-h we're both right, subsetting different things.  One OMC approval, one committer approval and one hold -- those are the "actionable" things.  Is this the discussion, or is it happening on the OMC list internally?

**Comment by richsalz** [comment]
I was mistaken, @FdaSilvaYY isn't a committer.  Glad the discussion, and eventual resolution I hope, is happening.

**Comment by FdaSilvaYY** [comment]
For a reason tat I ignore , I was able to approve. Sorry for the confusion.
GitHub issue, i suppose, 'cause I remember having the Comment choice enabled, in before days.

---
May I suggest adding a config option (yes, one more) to disable these old extensions support ?
Else it will take another decade before this removal happens.

**Comment by mspncp** [comment]
> For a reason that I ignore , I was able to approve. Sorry for the confusion.

It makes perfect sense that GitHub let's you approve or request changes. Because your opinion (and that of other contributors) is highly appreciated. The only difference is that your vote does not count pro or contra acceptance. If you look closely, you will notice that the check mark of your approval is colored grey, while Richard's check mark is green.

**Comment by richsalz** [comment]
This goes against the stated compatibility plans so I'm closing this. If someone wants to recreate for a future release, feel free to adopt the branch. :)

**Pull request closed without merging**